dilated-file-transfer.md

```diff
@@ -0,0 +1,476 @@
+
+Either side MAY send any number of Offer messages at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any offers.
+
+To make an offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+That is, a subchannel transmits a series of complete (framed) messages (up to ~4GiB in size).
+
+For this protocol, each record on the subchannel uses the first byte to indicate the kind of message; the remaining bytes in any message are kind-dependant.
+
+The following kinds of messages exist (indicated by the first byte):
+* 0x00: reserved / unused
+* 0x01: msgpack-encoded `FileOffer` message
+* 0x02: msgpack-encoded `DirectoryOffer` message
+* 0x03: msgpack-encoded `OfferAccept` message
+* 0x04: msgpack-encoded `OfferReejct` message
+* 0x05: file data bytes
+
+All other byte values are reserved for future use and MUST NOT be used.
+XXX: maybe spec [0, 128) as reserved, and [128, 255) for "experiments"?
+
+The first message sent on the new subchannel is either `FileOffer` or `DirectoryOffer`.
+
+To offer a single file (with message kind `1`):
+
+```python
+class FileOffer:
+    filename: str    # unicode relative pathname
+    timestamp: int   # Unix timestamp (seconds since the epoch in GMT)
+    bytes: int       # total number of bytes in the file
+```
+
+To offer a directory tree of many files (with message kind `2`):
+```python
+class DirectoryOffer:
+    base: str          # unicode path segment of the root directory (i.e. what the user selected)
+    size: int         # total number of bytes in _all_ files
+    files: list[str]   # a list containing relative paths for each file
+```
+
+The filenames in the `"files"` list are unicode relative paths (relative to the `"base"` from the `DirectoryOffer` and NOT including that part.
+
+For example:
+
+```python
+DirectoryOffer(
+    base="project",
+    size: 165,
+    files=["README", "src/hello.py"]
+    ]
+)
+```
+
+This indicates an offer to send two files, one in `"project/README"` and the other in `"project/src/hello.py"`.
+A client can consider the "base" name as suggestion, of course.
+On the flip side, a privacy-conscious sending application could offer to randomize the name when sending.
+
+In Haskell, this might look like:
+
+```haskell
+import Path (Dir, File, Rel, Path)
+import Data.Time (UTCTime)
+import Numeric.Natural (Natural)
+
+data Offer
+  = FileOffer {
+      fname :: Path Rel File,
+      timestamp :: UTCTime,
+      size :: Natural }
+  | DirectoryOffer{
+      base :: Path Rel Dir,
+      total_size :: Natural,
+      files :: [Path Rel File] }
+```
+
+
+What happens next depends on the mode of the peer.
+
+If the peer has `"mode": "yes"` then this peer MUST immediately start sending content messages (see below).
+
+If the peer has `"mode": "ask"` then this peer MUST NOT send any more messages and instead await an incoming message.
```

Comment:
Maybe I misunderstood your suggestion above -- I would leave _only_ "mode=ask" behavior, but maybe you meant to leave only "mode=yes" behavior?

Having "ask" as the default is predicated on preferring consent; that is, that a peer shouldn't have "wasted" a bunch of bandwidth (while e.g. waiting for their user to click "yes" or "no") since that's somewhat surprising.